lib/gemstash/gem_yanker.rb

```diff
@@ -0,0 +1,58 @@
+require "gemstash"
+
+module Gemstash
+  # Class that supports yanking a gem from the private repository of gems.
+  class GemYanker
+    include Gemstash::Env::Helper
+
+    # This error is thrown when yanking a non-existing gem name.
+    class UnknownGemError < StandardError
+    end
+
+    # This error is thrown when yanking a non-existing gem version.
+    class UnknownVersionError < StandardError
+    end
+
+    # This error is thrown when yanking an already yanked gem version.
+    class YankedVersionError < StandardError
+    end
+
+    def initialize(auth_key, gem_name, slug)
+      @auth_key = auth_key
+      @gem_name = gem_name
+      @slug = slug
+      @db_helper = Gemstash::DBHelper.new
+    end
+
+    def yank
+      check_auth
+      update_database
+      invalidate_cache
+    end
+
+  private
+
+    def check_auth
+      raise Gemstash::NotAuthorizedError, "Authorization key required" if @auth_key.to_s.strip.empty?
+      auth = Authorization[@auth_key]
+      raise Gemstash::NotAuthorizedError, "Authorization key is invalid" unless auth
+      raise Gemstash::NotAuthorizedError, "Authorization key doesn't have yank access" unless auth.yank?
```

Comment:
The checks are identical to those in `Gemstash::GemPusher`, so I moved it all to `Authorization.check(key, permission)` in 195aa822fce8f413c9427f460f3aa4507a2f559c. I couldn't think of a way to drop the line that is bothering you in the new `check` method though without combining both error cases into 1, which I don't want to do (the error message gets passed along as the `halt` message in the `Gemstash::GemSource::PrivateSource`, which is ultimately presented to the user by the `gem` command, so I think the distinction is useful)